### ISSUE TITLE
Does cyclonedds has program for gateway?

### ISSUE THREAD

**Issue by liyue75** [eclipse-cyclonedds/cyclonedds#584]
I'm now studing cyclonedds,  I want to know is there any part of this program specifically applied to gateways?
Is this program  just run on the client side?

**Comment by eboasson** [comment]
Hi @liyue75, Cyclone is a pure peer-to-peer system and there is no subset specifically for sensors or gateways. However, porting it in its entirety is pretty easy, all code that you need on the device is ordinary C code with all the (potentially) platform dependent stuff in the `ddsrt` component.

There are also some dependencies that you can ignore if you don't need some of the functionality. For example, you can just strip out all DDS Security support and support for TLS-secured TCP connections by setting a few build-time options (`ENABLE_SECURITY=NO` and `ENABLE_SSL=NO`). And even support for looking up hostnames is optional.

For a starting point you might want to take a look at the example port to FreeRTOS (https://github.com/eclipse-cyclonedds/cyclonedds/tree/master/ports/freertos-posix) — we know some people who are using Cyclone on an ARM Cortex M4 with a few MB of additional memory, using that port but with some adaptations to their specific hardware.

(Or my personal favourite: https://github.com/eclipse-cyclonedds/cyclonedds/tree/master/ports/solaris2.6 🙂)

Finally don't hesitate to ask for more information: your question was somewhat generic and so all I can do is try to point you in the right direction. I hope this is enough for you to decide to dig into it a bit further. I'm sure that will result in many more questions. Please, just ask them!

**Comment by liyue75** [comment]
Thanke you so much for your apply. 
I will go ahead to learn more about it. I want to port it to RIOS-OS with LWIP on Cortex M4. and I'll sahre the problems and results in the future.

**Comment by k0ekk0ek** [comment]
@liyue75, you might find this discussion (#481) helpful too. I outlined some of the steps there (we should create a guide at some point). There's also [this little document](https://github.com/eclipse-cyclonedds/cyclonedds/blob/master/docs/dev/freertos.md) that describes some of the requirements for FreeRTOS (but really embedded targets in general).

**Comment by liyue75** [comment]
Hi, I'm recently come back to study dds. I want to use it on riot-os(https://github.com/RIOT-OS/RIOT), it is an operation system for IoT which is similar to freertos, and can use lwip as network stack. 
After browing the documents, I find the easiest way to use cyclonedds on riot-os is to modify srcs in ddsrt module.
When using  FreeRTOS + LWIP, I find these srcs will be added to sources to compile in CMakeLists:
ddsrt                                       ddsrt                                                  ddsrt
 \- src                                         \- src                                                    \- src
      |- environ                                  |- random                                              |- string
      |       \- posix                             |         \- posix                                        |       \- posix
      |              \- environ.c                 |                 \- random.c                         |              \- strerror.c
      |- heap                                       |- rusage                                                |- sync
      |     \- freertos                            |       \- freertos                                      |       \- freertos
      |            \- heap.c                        |               \- rusage.c                            |               |- sync.c
      |- ifaddrs                                    |- sockets                                               |              \- tasklist.c
      |       \- lwip                                 |      \- posix                                          |- threads
      |             \- ifaddrs.c                     |             |- gethostname.c                    |        \- freertos
                                                          |             \- socket.c                              |                  \- thread.c  
                                                                                                                       |- time
                                                                                                                       |      \- freertos
                                                                                                                       |              \- time.c
 If I want to use RIOT + LWIP, it seems that I need to modify *.c files in the freertos folder above and in CMakeLists set macro WITH_FREERTOS and WITH_LWIP on. In freertos/*.c files, I need to replace the freertos's API with the one riot-os provides.
The problem now is that I don't konw much about freertos or cyclonedds. I don' t know what the appropriate API is for the substitution, and if just replacing the API will cause any problems. 
So I need to learn more about cyclonedds. But the source code is very large and it takes quite a bit time if I learn cyclonedds through the code. Is there a design document explaining how cyclonedds works or how can I quickly understand how it works?

**Comment by k0ekk0ek** [comment]
@liyue75, you're right in that you only need to make changes in `ddsrt`, that's where all the platform abstractions live. You probably want to use FreeRTOS (because it's the most basic OS we currently support, it's more like a task-scheduler actually) as an example. The [modules.md](https://github.com/eclipse-cyclonedds/cyclonedds/blob/master/docs/dev/modules.md) should provide you with a general idea of how the "modules" are stacked and the [freertos.md](https://github.com/eclipse-cyclonedds/cyclonedds/blob/master/docs/dev/freertos.md) file contains some info that might be useful. There's supposedly the native Linux port for RIOT OS. It sounds a bit like a simulation environment and that's probably a good place to start. You don't need to port everything, `ddsrt` is setup to be feature based, if a specific feature isn't offered by the platform, you can disable it in some cases. The lwIP part you can just reuse. The path I usually take is port everything required and then see if the `ddsrt` tests run. If that works, you're basically there.

**Comment by liyue75** [comment]
@k0ekk0ek, thanks, I 'll try port it this way.
There is one more question, are all documentation and instructions for cyclonedds in the docs folder? If I want to know how dds works, can I just read all the files in docs? Are there any other design documents or do I need to read the source code?

**Comment by liyue75** [comment]
@k0ekk0ek thank you for your reply. Do you know the minimum momery consuption of cyclonedds? If I want to use it on boards which has 1MB flash and 192KB ram, does it work?

**Comment by k0ekk0ek** [comment]
We definitely plan to run on very resource constrained devices, but I think this is too constrained. I got it to work on a Cortex-M4 once, but that device had 512KB SRAM and 16 MB SDRAM. If you do manage to get it working performance is going to be terrible since flash is extremely slow. Btw, if you do want to give it a try, you are better off getting started on somekind of simulator. That way you don't have to worry about writing linker scripts etc initially and just focus on the port, get that working and then worry about memory placement

**Comment by liyue75** [comment]
@k0ekk0ek Thaks a lot.

**Comment by zhangyaothu** [comment]
Hello @liyue75, do you have any advances? I am now trying to port cyclone dds on stm32h745 nucleo board with freertos and lwip. All tasks begin to run and SPDP begins to advertise, but malformed packets appear periodically as below  [rtps.zip](https://github.com/eclipse-cyclonedds/cyclonedds/files/7633178/rtps.zip). Do you have some advises?

**Comment by liyue75** [comment]
@zhangyaothu 
Hi, I didn't use cyclonedds last year because of memory problems with embedded devices.
Instead, I use Micro XRCE-DDS, [https://github.com/eProsima/Micro-XRCE-DDS](url), maybe you can try it.

**Comment by zhangyaothu** [comment]
Thanks @liyue75 , could you explain more about the restrictions? [TC39xx](https://www.infineon.com/cms/en/product/microcontroller/32-bit-tricore-microcontroller/32-bit-tricore-aurix-tc3xx/aurix-family-tc39xxx/) has 16MB FLASH and 6.9MB SRAM. This configuration may bring benifits.

**Comment by liyue75** [comment]
@zhangyaothu 
Sorry, I don't know. I didn't port cyclonedds after I knew it needed a lot of memory.
But Micro Xrce-DDS which implementing the DDS-XRCE protocol works well on resource constrained devices.

**Comment by zhangyaothu** [comment]
Thanks a lot @liyue75

**Comment by eboasson** [comment]
Can't help myself, I have to plug a lovely sibling Eclipse-project here: https://zenoh.io. @liyue75 if you have some spare time, you might to have a look at that, too, it is a far more elegant solution than XRCE 😀 There is a tiny implementation for very resource-constrained devices, it is very performant and wire-efficient and bridges transparently to DDS. In fact, the technology used for the [Indy Autonomous Challenge](https://www.indyautonomouschallenge.com) was Cyclone DDS in the car and Zenoh for V2X.

@zhangyaothu 16MB flash & 6.9MB SRAM is definitely sufficient for Cyclone. (Confirming it here in case anyone else finds this issue while looking for information on Cyclone and embedded devices.)

**Comment by liyue75** [comment]
@eboasson Thank you for your comment, I will take a look at that project.

**Comment by thijsmie** [comment]
I'm closing this because the discussion was concluded, relegated to knowledge base.

**Comment by sgf201** [comment]
> Can't help myself, I have to plug a lovely sibling Eclipse-project here: https://zenoh.io. @liyue75 if you have some spare time, you might to have a look at that, too, it is a far more elegant solution than XRCE 😀 There is a tiny implementation for very resource-constrained devices, it is very performant and wire-efficient and bridges transparently to DDS. In fact, the technology used for the [Indy Autonomous Challenge](https://www.indyautonomouschallenge.com) was Cyclone DDS in the car and Zenoh for V2X.
> 
> @zhangyaothu 16MB flash & 6.9MB SRAM is definitely sufficient for Cyclone. (Confirming it here in case anyone else finds this issue while looking for information on Cyclone and embedded devices.)

after took a  look at zenoh.io, are you sure it is suitable for mcu? I can not find any info about rtos in the issues

**Comment by eboasson** [comment]
Certainly https://github.com/eclipse-zenoh/zenoh-pico fits on any MCU and runs on various RTOSs. I think the best thing to is to ping them on GitHub or Discord. Perhaps @Mallets, @JEnoch or any of the others working on Zenoh is willing to give some information here as well.

**Comment by JEnoch** [comment]
Hi @sgf201,
You can have a look to this blog we posted last year:
https://zenoh.io/blog/2021-11-09-ros2-zenoh-pico/
Here, we show how [zenoh-pico](https://github.com/eclipse-zenoh/zenoh-pico) runs on a ESP32 MCU, and can publish/subscribe to DDS with the [zenoh-bridge-dds](https://github.com/eclipse-zenoh/zenoh-plugin-dds) as intermediate.
If you have more questions, join us on Discord: https://discord.gg/vSDSpqnbkm !
